[macOS] Add support for theming native controls using NSViewSemanticContext

pxlcoder opened the pull request:
#### 174814803dcd46a72695730e8994155008f0e034
<pre>
[macOS] Add support for theming native controls using NSViewSemanticContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=248764">https://bugs.webkit.org/show_bug.cgi?id=248764</a>
rdar://99309431

Reviewed by Wenson Hsieh and Tim Horton.

Allow clients to customize the appearance of controls using the web view’s
semantic context. This functionality is achieved by adding support for
`-[WKWebView _setSemanticContext]`, which is existing `NSView` SPI.

Support is restricted to `NSViewSemanticContextForm`, as other contexts do not
apply to web views.

* Source/WTF/wtf/PlatformUse.h:

Add a USE() macro rather than a HAVE() macro, since `NSViewSemanticContextForm`
is only available on macOS 13, even though `NSViewSemanticContext` is available
on older macOS. Consequently, the entire `NSViewSemanticContext` is only being
*used* on macOS 13, as WKWebView support is restricted to form semantic contexts.

* Source/WebCore/PAL/pal/spi/mac/NSViewSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::ThemeMac::ensuredView):
(WebCore::ThemeMac::setUseFormSemanticContext):
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/mac/WKView.mm:
(-[WKView _setSemanticContext:]):
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _setSemanticContext:]):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::useFormSemanticContext const):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:

Remove a redundant `#if PLATFORM(MAC)`.

(WebKit::WebPageProxy::useFormSemanticContext const):
(WebKit::WebPageProxy::semanticContextDidChange):
(WebKit::WebPageProxy::colorSpace):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::useFormSemanticContext const):
(WebKit::WebViewImpl::semanticContextDidChange):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::WebPage):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::setUseFormSemanticContext):
(WebKit::WebPage::semanticContextDidChange):

Canonical link: <a href="https://commits.webkit.org/257792@main">https://commits.webkit.org/257792@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b86035afa8ab0f62ffbdf71837ab98af960fe7f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99926 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9095 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33005 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109276 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169513 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9976 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86389 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92374 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107179 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105693 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7530 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90820 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34255 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22211 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90544 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2896 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23724 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86460 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2853 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/46093 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29002 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5350 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8985 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43198 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89340 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4704 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19974 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/platform/mac/ThemeMac.mm:
```diff
@@ -42,12 +42,14 @@
 
 static NSRect focusRingClipRect;
 static BOOL themeWindowHasKeyAppearance;
+static bool useFormSemanticContext = false;
```
No need for `= false`

hortont424 replied:
Also, should we pick BOOL or bool for both?

pxlcoder replied:
`themeWindowHasKeyAppearance` is returned directly from a `BOOL` method, whereas `useFormSemanticContext` is plumbed in as a `bool`.

Do you think I should change `useFormSemanticContext` to a `BOOL`?

hortont424 commented on Source/WebCore/PAL/pal/spi/mac/NSViewSPI.h:
```diff
@@ -29,11 +29,39 @@
+#if USE(NSVIEW_SEMANTICCONTEXT)
+
+typedef NS_ENUM(NSInteger, NSViewSemanticContext) {
+    NSViewSemanticContextInherit,
```
Could just mention the one you need by making its value explicit, instead of exposing all this SPI

pxlcoder committed:
[macOS] Add support for theming native controls using NSViewSemanticContext
https://bugs.webkit.org/show_bug.cgi?id=248764
rdar://99309431

Reviewed by Wenson Hsieh and Tim Horton.

Allow clients to customize the appearance of controls using the web view’s
semantic context. This functionality is achieved by adding support for
`-[WKWebView _setSemanticContext]`, which is existing `NSView` SPI.

Support is restricted to `NSViewSemanticContextForm`, as other contexts do not
apply to web views.

* Source/WTF/wtf/PlatformUse.h:

Add a USE() macro rather than a HAVE() macro, since `NSViewSemanticContextForm`
is only available on macOS 13, even though `NSViewSemanticContext` is available
on older macOS. Consequently, the entire `NSViewSemanticContext` is only being
*used* on macOS 13, as WKWebView support is restricted to form semantic contexts.

* Source/WebCore/PAL/pal/spi/mac/NSViewSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::ThemeMac::ensuredView):
(WebCore::ThemeMac::setUseFormSemanticContext):
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/mac/WKView.mm:
(-[WKView _setSemanticContext:]):
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _setSemanticContext:]):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::useFormSemanticContext const):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:

Remove a redundant `#if PLATFORM(MAC)`.

(WebKit::WebPageProxy::useFormSemanticContext const):
(WebKit::WebPageProxy::semanticContextDidChange):
(WebKit::WebPageProxy::colorSpace):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::useFormSemanticContext const):
(WebKit::WebViewImpl::semanticContextDidChange):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::WebPage):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::setUseFormSemanticContext):
(WebKit::WebPage::semanticContextDidChange):

Canonical link: https://commits.webkit.org/257792@main
